### PR TITLE
[nrf fromtree] Image access hooks sandbox

### PR DESCRIPTION
**Pull request by nvlsianpu**
This PR is for shows idea of having hook for override MCUboot's internal procedures for access an image.
Objective is to be able to run upstream multi-image feature on nRF53 device.

- [x] Kconfig for hooks
- [x] hooks needed for chainn-load application
- [x] hooks needed for update application
- [x] hooks needed for serial recovery

Currently build can be tested on nrf52840dk:
You might use configuration sample form this helper-draft https://github.com/mcu-tools/mcuboot/pull/1084
For building mcuboot use following comand-line:
``` console
west build -d build/mcuboot_y -b nrf52840dk_nrf52840 \
bootloader/mcuboot/boot/zephyr/ -- \
-DDTC_OVERLAY_FILE=./boards/nrf52840dk_ram_ext.overlay \
-DOVERLAY_CONFIG=./boards/flash_ext_sim_driver.conf
```
or

``` console
west build -d build/mcuboot_y -b nrf52840dk_nrf52840 \
bootloader/mcuboot/boot/zephyr/ -- \
-DDTC_OVERLAY_FILE=./boards/nrf52840dk_ram_multi.overlay \
-DOVERLAY_CONFIG=./boards/flash_ext_sim_driver.conf
```
mcuboot will  be able to chainn-load signed application from the firs image-pair primary slot

parent upstream PR: https://github.com/mcu-tools/mcuboot/pull/1058

### CODE REVIEW

File: boot/bootutil/src/loader.c [hunk 1]
```diff
@@ -78,6 +78,34 @@ static struct boot_loader_state boot_data;
 #define TARGET_STATIC
 #endif
 
+/* @retval 0: header was read/populated
+ *         FIH_FAILURE: image is invalid,
+ *         1 if hook not implemented for the image-slot,
+ *         othervise an error-code value.
+ */
+int boot_read_image_header_hook(int img_index, int slot, struct image_header *img_hed)
```
Comment (nvlsianpu) [inline boot/bootutil/src/loader.c:86]: simple implementation is here only for show it

File: boot/bootutil/src/loader.c [hunk 2]
```diff
@@ -78,6 +78,34 @@ static struct boot_loader_state boot_data;
...
+    }
+
+    return 1;
+}
+
+/* @retval FIH_SUCCESS: image is valid,
+ *         FIH_FAILURE: image is invalid,
+ *         fih encoded 1 if hook not implemented for the image-slot.
+ */
+fih_int boot_image_check_hook(int img_index)
```
Comment (nvlsianpu) [inline boot/bootutil/src/loader.c:100]: simple implementation is here only for show it

File: boot/bootutil/src/loader.c [hunk 3]
```diff
@@ -1648,7 +1662,7 @@ boot_prepare_image_for_update(struct boot_loader_state *state,
     rc = boot_read_image_headers(state, false, NULL);
     if (rc != 0) {
         /* Continue with next image if there is one. */
-        BOOT_LOG_WRN("Failed reading image headers; Image=%u",
+        BOOT_LOG_WRN("Failed reading image headers; lol Image=%u",
```
Comment (sigvartmh) [inline boot/bootutil/src/loader.c:1665]: Guess this is not needed

File: boot/zephyr/hooks_sample.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        FIH_RET(FIH_SUCCESS);
+    }
+
+    FIH_RET(fih_int_encode(1));
+}
+
+int boot_perform_update_hook(int img_index, struct image_header *img_head,
+                             const struct flash_area *area)
+{
+    if (img_index == 1) {
```
Comment (sigvartmh) [inline boot/zephyr/hooks_sample.c:55]: So do you think we should add any information here to describe where the image is going to or is the slot numbering enough?

So an example for nrf53 we would have
```
image_index == 0
                rc = boot_perform_update(state, &bs);
image_index == 1
               rc = pcd_update(...)

File: boot/bootutil/src/loader.c (hunk 3 above)
Comment (nvlsianpu) [reply boot/bootutil/src/loader.c:1665]: debugging artefact 😉 : , will remove

File: boot/zephyr/hooks_sample.c (hunk 4 above)
Comment (MarekPieta) [reply boot/zephyr/hooks_sample.c:55]: Maybe we could simply define names for the constant here?

Comment (nvlsianpu) [reply boot/zephyr/hooks_sample.c:55]: @sigvartmh Idea is as folow:
``` C
image_index == 0
                return 1; /* boot_perform_update() will be cale in normal code flow */
image_index == 1
               rc = pcd_update(...)
               ...
```
Actually i found better place to intoduce hook for call `pcd_update(...)` -> see `int boot_copy_region_post_hook()`

File: boot/bootutil/include/bootutil/boot_hooks.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ * implemented to overide or to amend some of MCUboot's native routines.
+ */
+
+#ifndef H_BOOTUTIL_HOOKS
+#define H_BOOTUTIL_HOOKS
+
+/** Hook for provide image header data.
+ *
+ * This Hook may be used to overide image header read implementation or doing
+ * a custom action befor.
```
Comment (de-nordic) [inline boot/bootutil/include/bootutil/boot_hooks.h:40]: s/befor/before/

**Comment by nvlsianpu** [comment]
Updated

File: boot/zephyr/hooks_sample.c [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    return BOOT_HOOK_REGULAR;
+}
+
+int boot_read_swap_state_primary_slot_hook(int image_index,
+                                           struct boot_swap_state *state)
+{
+    if (image_index == 1) {
+        state->magic = BOOT_MAGIC_UNSET;
+        state->swap_type = BOOT_SWAP_TYPE_NONE;
+        state->image_num = image_index ; // ?
```
Comment (de-nordic) [inline boot/zephyr/hooks_sample.c:70]: some leftover?

File: boot/zephyr/hooks_sample.c [hunk 7]
```diff
@@ -0,0 +1,90 @@
+/*
```
Comment (hakonfam) [inline boot/zephyr/hooks_sample.c:1]: Is this code built and executed by any samples? If so should that be included in this commit? Or is the sample in another repo?

Comment (hakonfam) [reply boot/zephyr/hooks_sample.c:1]: I see now that you are linking to a different PR in the PR description which uses this file, would it be possible to include those changes in this PR as well? To me it makes little sense to commit a file which is not included in any build.

File: boot/bootutil/src/bootutil_public.c [hunk 8]
```diff
@@ -417,8 +419,14 @@ boot_swap_type_multi(int image_index)
...
     size_t i;
 
-    rc = boot_read_swap_state_by_id(FLASH_AREA_IMAGE_PRIMARY(image_index),
-                                    &primary_slot);
+#ifdef MCUBOOT_IMAGE_ACCESS_HOOKS
+    rc = boot_read_swap_state_primary_slot_hook(image_index, &primary_slot);
+    if (rc == BOOT_HOOK_REGULAR)
+#endif
+    {
+        rc = boot_read_swap_state_by_id(FLASH_AREA_IMAGE_PRIMARY(image_index),
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/bootutil/src/bootutil_public.c:427]: Yes it is intended:
- this hook covers primary_slot state
- this patch covers MCUboot's overwrite only mode

File: boot/bootutil/include/bootutil/boot_hooks.h [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+#ifndef H_BOOTUTIL_HOOKS
+#define H_BOOTUTIL_HOOKS
+
+/** Hook for provide image header data.
+ *
+ * This Hook may be used to overide image header read implementation or doing
+ * a custom action before.
+ *
+ * @param img_index the index of the images pair
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/bootutil/include/bootutil/boot_hooks.h:42]: images pair-> primary and secondary slot of given image_index

File: boot/zephyr/Kconfig [hunk 10]
```diff
@@ -649,6 +649,20 @@ config BOOT_WATCHDOG_FEED
...
+	help
+	  Allow to provide procedures for override native MCUboot's routines
+	  required for access the image data and the image update.
+
+config BOOT_IMAGE_ACCESS_HOOKS_FILE
+	string "Hooks implementation file path"
+	depends on BOOT_IMAGE_ACCESS_HOOKS
+	default ""
+	help
+	  Path to the file which implements hooks. Path is relative to
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/zephyr/Kconfig:663]: good point

File: boot/zephyr/include/mcuboot_config/mcuboot_config.h [hunk 11]
```diff
@@ -144,6 +144,16 @@
 #define MCUBOOT_FIH_PROFILE_HIGH
 #endif
 
+#ifdef CONFIG_ENABLE_MGMT_PERUSER
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/zephyr/include/mcuboot_config/mcuboot_config.h:147]: ops

File: boot/zephyr/Kconfig [hunk 12]
```diff
@@ -649,6 +649,20 @@ config BOOT_WATCHDOG_FEED
 	  Enables implementation of MCUBOOT_WATCHDOG_FEED() macro which is
 	  used to feed watchdog while doing time consuming operations.
 
+config BOOT_IMAGE_ACCESS_HOOKS
+	bool "Hooks for override MCUboot's native routines"
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/zephyr/Kconfig:653]: Done

File: boot/zephyr/Kconfig (hunk 10 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/zephyr/Kconfig]: Done

File: boot/zephyr/include/mcuboot_config/mcuboot_config.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/zephyr/include/mcuboot_config/mcuboot_config.h]: removed

File: boot/zephyr/hooks_sample.c (hunk 7 above)
Comment (nvlsianpu) [reply boot/zephyr/hooks_sample.c:1]: I added test-build in sample.yaml

File: boot/bootutil/include/bootutil/boot_hooks.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boot/bootutil/include/bootutil/boot_hooks.h]: In other file `image pair` term is used. I corrected to this term here

**Comment by nvlsianpu** [comment]
@hakonfam Updated

File: boot/zephyr/CMakeLists.txt [hunk 13]
```diff
@@ -309,3 +309,27 @@ zephyr_library_sources(
...
+    set(HOOKS_FILE ${BOOT_DIR}/zephyr/${CONFIG_BOOT_IMAGE_ACCESS_HOOKS_FILE})
+  endif()
+
+
+  if(DEFINED HOOKS_FILE)
+    zephyr_library_sources(
+      ${HOOKS_FILE}
+    )
+  else()
+    message(FATAL_ERROR "Can't access hooks implementation file.")
```
Comment (hakonfam) [inline boot/zephyr/CMakeLists.txt:332]:
```suggestion
    message(WARNING "Can't access hooks implementation file.")
```

This way cmake can finish and the users are able to use `ninja menuconfig` or similar to fix the mis-configuration.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: boot/zephyr/Kconfig [hunk 14]
```diff
@@ -649,6 +649,25 @@ config BOOT_WATCHDOG_FEED
...
+	bool "Enable hooks for overriding MCUboot's native routines"
+	help
+	  Allow to provide procedures for override or extend native
+	  MCUboot's routines required for access the image data and the image
+	  update.
+
+config BOOT_IMAGE_ACCESS_HOOKS_FILE
+	string "Hooks implementation file path"
+	depends on BOOT_IMAGE_ACCESS_HOOKS
+	default ""
```
Comment (hakonfam) [inline boot/zephyr/Kconfig:662]:
```suggestion
```

Remove this so that those that include this as a module can define a different default for it.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: boot/zephyr/CMakeLists.txt (hunk 13 above)
Comment (nvlsianpu) [reply boot/zephyr/CMakeLists.txt]: done

File: boot/zephyr/Kconfig (hunk 14 above)
Comment (nvlsianpu) [reply boot/zephyr/Kconfig]: done

**Review by hakonfam — APPROVED** [review]

File: boot/bootutil/include/bootutil/boot_hooks.h [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+int boot_copy_region_post_hook(int img_index, const struct flash_area *area,
+                               size_t size);
+
+/** Hook for implement image's post recovery upload action
+ *
+ * This hook is for implement action which might be done right after image was
+ * copied to the primary slot. This hook is called in serial recovery upload
+ * operation.
+ *
+ * @param img_index the index of the image pair
```
Comment (hakonfam) [inline boot/bootutil/include/bootutil/boot_hooks.h:128]: Is this correct? To me this seems to be the slot index.

Comment (nvlsianpu) [reply boot/bootutil/include/bootutil/boot_hooks.h:128]: This is correct - this is hook called after image get uploaded completely in serial recovery operation. Upload is always do to the primary slot of given img_index.

**Comment by nvlsianpu** [comment]
CI on the Manifest PR passed, so I'm merging this. (https://github.com/nrfconnect/sdk-nrf/pull/5359)

**Pull request merged**